sdk/data/azcosmos/cosmos_client_retry_policy.go

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azcosmos
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	azruntime "github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/internal/errorinfo"
+)
+
+type clientRetryPolicy struct {
+	gem                    *globalEndpointManager
+	useWriteEndpoint       bool
+	retryCount             int
+	sessionRetryCount      int
+	preferredLocationIndex int
+}
+
+const maxRetryCount = 120
+const defaultBackoff = 1
+
+func (p *clientRetryPolicy) Do(req *policy.Request) (*http.Response, error) {
+	p.resetPolicyCounters()
+	o := pipelineRequestOptions{}
+	if !req.OperationValue(&o) {
+		return nil, fmt.Errorf("failed to obtain request options, please check request being sent: %s", req.Body())
+	}
+	for {
+		resolvedEndpoint := p.gem.ResolveServiceEndpoint(p.retryCount, o.isWriteOperation, p.useWriteEndpoint)
+		req.Raw().Host = resolvedEndpoint.Host
+		req.Raw().URL.Host = resolvedEndpoint.Host
+		response, err := req.Next() // err can happen in weird scenarios (connectivity, etc)
+		if err != nil {
+			if p.isNetworkConnectionError(err) {
+				shouldRetry, err := p.attemptRetryOnNetworkError(req)
+				if err != nil {
+					return nil, err
+				}
+				if !shouldRetry {
+					return nil, errorinfo.NonRetriableError(azruntime.NewResponseErrorWithErrorCode(response, response.Status))
```

Comment:
This is a bug I left, sorry!

```
shouldRetry, errRetry := p.attemptRetryOnNetworkError(req)
if errRetry != nil {
	return nil, errRetry
}
if !shouldRetry {
	return nil, err
```

If shouldRetry is false, we should return the original error